compaction/compaction.cc

```diff
@@ -759,14 +760,9 @@ class compaction {
             // sstable than just adding up the lengths of individual sstables.
             _estimated_partitions += sst->get_estimated_key_count();
             _compacting_data_file_size += sst->ondisk_data_size();
-            // TODO:
-            // Note that this is not fully correct. Since we might be merging sstables that originated on
-            // another shard (#cpu changed), we might be comparing RP:s with differing shard ids,
-            // which might vary in "comparable" size quite a bit. However, since the worst that happens
-            // is that we might miss a high water mark for the commit log replayer,
-            // this is kind of ok, esp. since we will hopefully not be trying to recover based on
-            // compacted sstables anyway (CL should be clean by then).
-            _rp = std::max(_rp, sst_stats.position);
+            if (sst->originated_on_this_node().value_or(true) && sst_stats.position.shard_id() == this_shard_id()) {
```

Comment:
Should it be value_or(false)? If the sstable doesn't have a host id, could it have come from this node?

What about an sstable that migrated to another node, and then back?

Reply:
It is disengaged only for old sstables with old version.
For backward compatibility we can assume they originated on this node.

As for migrating back and forth, does it really need to invalidate the replay position?
For tablets, I think we need not support rebuild node operation, but rather for the rebuild use case, we can bootstrap the node as a new one and use removenode to rebuild the the tablets it used to own using the new node incarnation and possible existing nodes.

Reply:
I'll change the default to `false` in the next version so to use the replay_position only if it is valid for sure.
We don't have the local host_id yet when resharding the system tables, and in the resharding case the replay position of the output table cannot be derived from the input sstables anyhow in the typical case where the output sstables land on a different shard than the input sstables.

Reply:
I'm going back one step, what does the calculation mean? what does std::max(replay_position in sstables) give us? we know anything after that must be replayed, but things before that point also need to be replayed, no?

Reply:
The maximum calculated here, in compaction, is used for setting the replay_position in the output sstable compaction is writing.

> but things before that point also need to be replayed, no?

Memtable flushes are serialized so why things below the highest replay position need to be replayed?

Reply:
Per table / compaction group, that is.
Things may need to be replayed for other tables that did not flush their memtables yet.

Reply:
> The maximum calculated here, in compaction, is used for setting the replay_position in the output sstable compaction is writing.
> 
> > but things before that point also need to be replayed, no?
> 
> Memtable flushes are serialized so why things below the highest replay position need to be replayed?



Ok, I see - we want to replay from max(replay position) and to be able to calculate it after compaction we apply max() to the inputs.

Reply:
But with tablets memtable flushes are not serialized. So either we track replay_position per tablet or we use min(max(replay_position in tablet) across all tablets)